### PR TITLE
dev: Set up Apollo Client for local state management

### LINKED ISSUE #208: Custom links/collections CRUD (local only, no DB)
Set up a data model and CRUD actions that interact with local state (React context, local storage, etc.), with the intention of swapping out for a NoSQL database at a later point.

proposed types:
```
Bookmark {
  url: string
  label?: string
  description?: string
}

Collection {
  title: string
  bookmarks: Bookmark[]
}
```

Actions:
- Add a bookmark (to a collection)
- Delete a bookmark

This issue covers user-generated bookmarks and collections (not CMS).

Epics: 
https://github.com/USSF-ORBIT/ussf-portal-client/issues/153
https://github.com/USSF-ORBIT/ussf-portal-client/issues/154
https://github.com/USSF-ORBIT/ussf-portal-client/issues/179
https://github.com/USSF-ORBIT/ussf-portal-client/issues/181

### PR DESCRIPTION
**Pull request by abbyoung**
## Description 

Sets up Apollo Client and uses local resolvers to mimic db operations with the cache.

* Adds operations for adding/removing bookmarks and collections
* Adds local resolvers to store this data in the cache
* Persists cache data using local storage

Fixes #208 

## Review Notes
Tests to be added once we write components that use these operations.

### CODE REVIEW

File: src/pages/_app.tsx [hunk 1]
```diff
@@ -4,14 +4,46 @@ import type { AppProps } from 'next/app'
...
+  return cache
+}
+
+// Set up client with persisted cache and local resolvers
+const client = new ApolloClient({
+  cache: initCache(),
+  resolvers: localResolvers,
+})
+
+/* End ApolloClient setup */
```
Comment (suzubara) [inline src/pages/_app.tsx:46]: ☝️  Can we move the above setup to a separate file (like `src/stores/apolloClient.ts`) that exports the `client` object?

**Review by suzubara — COMMENTED** [review]
I left a few minor things, I am also questioning whether it might still be useful to define the client-side schema even if it is not used for validation, just to document how we are intending to use these types?

File: package.json [hunk 2]
```diff
@@ -74,6 +77,7 @@
     "eslint-plugin-react": "7.25.1",
     "eslint-plugin-react-hooks": "4.2.0",
     "eslint-plugin-security": "1.4.0",
+    "graphql": "^15.5.3",
```
[the remark this answers is not in the document]
Comment (abbyoung) [reply package.json:80]: thank you!

File: src/operations/mutations/addCollection.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+interface AddCollectionInput {
+  id: string
+  title: string
+  bookmarks: Bookmark[]
+}
+
+export const ADD_COLLECTION = gql`
+  mutation addCollection($id: ID, $title: String, $bookmarks: [Bookmark]) {
+    addCollection(id: $id, title: $title, bookmarks: $bookmarks) @client
```
[the remark this answers is not in the document]
Comment (abbyoung) [reply src/operations/mutations/addCollection.tsx:18]: I had given it a try and kept running into problems, but let me try again and then I can at least give you a better answer 😄 and most likely fix it

File: src/operations/resolvers.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            : [newCollection],
+      }
+
+      cache.writeQuery({ query: GET_COLLECTIONS, data })
+      return newCollection
+    },
+    removeCollection: (_root, { id }, { cache }) => {
+      cache.modify({
+        fields: {
+          collections(collections: Collection[], { readField }: any) {
```
Comment (abbyoung) [inline src/operations/resolvers.tsx:68]: I need to figure out what type `readField` is

File: src/operations/resolvers.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
+import { Resolvers } from '@apollo/client'
+import { v4 } from 'uuid'
+
+import { GET_COLLECTIONS } from './queries/getCollections'
+import type { Bookmark, Collection } from 'types'
+
+export const localResolvers: Resolvers | Resolvers[] = {
+  Query: {
+    collections: (_root, { id }, { cache }) => {
```
Comment (abbyoung) [inline src/operations/resolvers.tsx:9]: @suzubara this resolver has my brain in a STATE and i'd love to pair a bit to talk through this. 

- it's not always triggered  when i would expect it to
- it seems to have no effect on the rendering of my MySpace component in another branch
- it _is_ required to query as expected in GraphiQL

File: src/operations/mutations/addCollection.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (abbyoung) [reply src/operations/mutations/addCollection.tsx]: resolved 🌟

File: src/operations/resolvers.tsx (hunk 5 above)
Comment (abbyoung) [reply src/operations/resolvers.tsx:9]: Discussed with @suzubara -- leaving in for now and revisiting later

File: src/operations/resolvers.tsx (hunk 4 above)
Comment (abbyoung) [reply src/operations/resolvers.tsx:67]: passing for now but suz found this resource ffr: https://www.apollographql.com/blog/apollo-client/architecture/redux-to-apollo-data-access-patterns/

**Review by suzubara — APPROVED** [review]
✨ let's do it! we haven't been able to validate using apollo devtools but that is most likely because of this issue: https://github.com/apollographql/apollo-client-devtools/issues/239
